src/lib/network_pool/test/indexed_pool_tests.ml

```diff
@@ -920,3 +920,122 @@ let application_invalidates_applied_transactions () =
+        |> Result.map_error ~f:(fun e ->
+               Sexp.to_string @@ Command_error.sexp_of_t e )
+        |> Result.ok_or_failwith
+      in
+      [%test_eq: int] (Indexed_pool.size pool) (Indexed_pool.size pool') ;
+      [%test_eq: int]
+        (Indexed_pool.transactions ~logger pool |> Sequence.length)
+        (Indexed_pool.transactions ~logger pool' |> Sequence.length) )
+
+let transaction_replacement_insufficient_balance () =
+  Quickcheck.test
+    (let open Quickcheck.Generator.Let_syntax in
+    let%bind a = Account.gen in
+    let sender = { a with balance = Balance.of_mina_int_exn 31 } in
+    let%map recv = Account.gen in
+    let cmds =
+      List.init 3 ~f:(fun n ->
+          let open Signed_command.Payload in
+          let (`If_this_is_used_it_should_have_a_comment_justifying_it cmd) =
+            Signed_command.Poly.
+              { payload =
+                  Poly.
+                    { common =
+                        Common.Poly.
+                          { fee =
+                              Fee.of_nanomina_int_exn 300_000_000 (* 0.3 Mina *)
+                          ; fee_payer_pk = sender.public_key
+                          ; nonce = Account_nonce.(add sender.nonce @@ of_int n)
+                          ; valid_until = Global_slot_since_genesis.max_value
+                          ; memo = Signed_command_memo.dummy
+                          }
+                    ; body =
+                        Body.Payment
+                          Payment_payload.Poly.
+                            { receiver_pk = recv.public_key
+                            ; amount = Amount.of_mina_int_exn 10
+                            }
+                    }
+              ; signer =
+                  Option.value_exn @@ Public_key.decompress sender.public_key
+              ; signature = Signature.dummy
+              }
+            |> Signed_command.to_valid_unsafe
+          in
+          cmd )
+    in
+    (sender, cmds))
+    ~f:(fun (sender, txns) ->
+      let account_map = accounts_map [ sender ] in
+      let pool = pool_of_transactions ~account_map txns in
+      let t = List.nth_exn txns 1 |> Signed_command.forget_check in
+      let updated =
+        update_fee t (Currency.Fee.of_nanomina_int_exn 700_000_000)
+        (* 0.7 Mina *)
+      in
+      let (`If_this_is_used_it_should_have_a_comment_justifying_it replacement)
+          =
+        Signed_command.to_valid_unsafe updated
+      in
+      let t =
+        User_command.Signed_command replacement
+        |> Transaction_hash.User_command_with_valid_signature.create
+      in
+      let balance = Balance.to_amount sender.balance in
+      let _, pool', _ =
+        Indexed_pool.add_from_gossip_exn pool t sender.nonce balance
+        |> Result.map_error ~f:(fun e ->
+               Sexp.to_string @@ Command_error.sexp_of_t e )
+        |> Result.ok_or_failwith
+      in
+      (* The last transaction gets discarded, because after the replacement,
```

Comment:
Didn't fully follow the logic here: is it discarded because of higher fee or higher amount?